Basic TensorFlow implementation of EM, and associated changes.

andychu opened the pull request:
It can be selected by passing --em-executable as fast_em.sh (wrapper for
Python) instead of the binary compiled from fast_em.cc.

See comments at the top of fast_em.py.

- association_test.R: test all three implementations
- bin/test.sh
  - Add demo for TensorFlow
  - Add demo of early convergence
  - Put the output from each implementation in its own directory.
- fast_em.R: Test exit code when shelling out to --em-executable
- Write number of EM iterations in the assoc-metrics.json output (all
  implementations).  This changes the protocol between the R driver and
  EM implementation.
- Clean up the log output from fast_em.cc

tkaitchuck commented on analysis/tensorflow/fast_em.py:
```diff
@@ -0,0 +1,169 @@
+  # split along dimension 0
+  v_split = tf.split(0, num_entries, v)
+
+  z_numerator = [report * pij_in for report in v_split]
```
Break this out into a function for clarity.

tkaitchuck commented on analysis/tensorflow/fast_em.py:
```diff
@@ -0,0 +1,169 @@
+  pij_in = tf.placeholder(tf.float64, shape=(entry_size,))
+
+  # split along dimension 0
+  v_split = tf.split(0, num_entries, v)
```
See if there is a better way to do this. Either through a different OP as Olfar suggested or by reusing a constant v_split.

tkaitchuck commented on analysis/tensorflow/fast_em.sh:
```diff
@@ -0,0 +1,23 @@
+  export CUDA_HOME=/usr/local/cuda-7.0
+
+  # Never returns
+  exec $THIS_DIR/fast_em.py "$@"
```
You shouldn't clobber the existing variable for LD_LIBRARY_PATH rather you should append that to it. 
Also you don't need to export, you can just assign the variables on the same line as the exec and the child will inherent them without messing with them in the user's shell afterwards.

tkaitchuck commented:
Aside from comments above, looks good. Merge when ready.

andychu commented:
OK, I addressed all the comments, except I put a TODO for the v_split thing.  I agree that is probably the issue, but I'm not sure how to fix it yet and I have to do some other stuff.  Right now it is over 200x slower!
